### PR TITLE
EZP-23924: Empty sort_key_string should be fixed to prevent failures

### PR DESCRIPTION
**Pull request by pspanja**
This PR resolves https://jira.ez.no/browse/EZP-23924

When fixing sort key field type bugs we did not always provide update scripts, leaving the database in inconsistent state. This could result in incorrect matches when using search.

This implements the cleanup script that will check if fields of given field type have correct sort key, and update it if needed.

### CODE REVIEW

File: eZ/Bundle/EzPublishMigrationBundle/Command/LegacyStorage/UpdateSortKeysCommand.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                        $query->bindValue( "-", null, PDO::PARAM_STR ),
+                        $databaseHandler->quoteColumn( "version" )
+                    )
+                )
+            )
+            ->from( "ezcontentobject_attribute" )
+            ->where(
+                $query->expr->eq(
+                    $databaseHandler->quoteColumn( "data_type_string" ),
+                    $query->bindValue( $fieldTypeIdentifier )
```
Comment (glye) [inline eZ/Bundle/EzPublishMigrationBundle/Command/LegacyStorage/UpdateSortKeysCommand.php]: Does this fetch all rows from ezcontentobject_attribute of the given data type string, without limits? If so, could be a problem. Tends to be crowded, that table.

File: eZ/Bundle/EzPublishMigrationBundle/Command/LegacyStorage/UpdateSortKeysCommand.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        }
+
+        $query = $databaseHandler->createSelectQuery();
+        $query
+            ->selectDistinct( "contentobject_id, version" )
+            ->from( "ezcontentobject_attribute" )
+            ->where(
+                $query->expr->eq(
+                    $databaseHandler->quoteColumn( "data_type_string" ),
+                    $query->bindValue( $fieldTypeIdentifier )
```
Comment (glye) [inline eZ/Bundle/EzPublishMigrationBundle/Command/LegacyStorage/UpdateSortKeysCommand.php]: Does this fetch all rows from ezcontentobject_attribute of the given data type string, without limits? If so, could be a problem. Tends to be crowded, that table.

Comment (pspanja) [inline eZ/Bundle/EzPublishMigrationBundle/Command/LegacyStorage/UpdateSortKeysCommand.php]: Not without limits, see `loadData()` method.

**Comment by glye** [comment]
+1

**Comment by yannickroger** [comment]
Do we need to create a dedicated bundle to host these scripts ?
Also don't we need BC doc to tell people to run it ?

**Comment by andrerom** [comment]
+1 I guess migration bundle can make sense down the road, but as gl asks, any particular reason to introduce it now already?

**Comment by pspanja** [comment]
@yannickroger 

> Also don't we need BC doc to tell people to run it ?

Added upgrade notes for 6.0 in https://github.com/ezsystems/ezpublish-kernel/commit/82c14f5890f401356a788acd24740d055142c475, anticipating this won't get into 5.3.5 and 5.4.2. I will add notes for 5.3 and 5.4 with backports.

Upgrade should cover what was previously reported in https://jira.ez.no/browse/EZP-23894 and https://jira.ez.no/browse/EZP-23894, so it instructs updating `ezstring` and `ezobjectrelation` type fields.

@andrerom 

>  I guess migration bundle can make sense down the road, but as gl asks, any particular reason to introduce it now already?

Not really, MigrationBundle was mentioned once before and I had to put it somewhere. This open for better suggestion of course.

**Comment by pspanja** [comment]
There is one issue here, the script will default to using repository configured for the default siteaccess, so it can't handle multirepo installations nicely. I tried to implement siteaccess selection through request context, but with no luck.

@lolautruche: is it possible to do this at all ATM?

**Comment by yannickroger** [comment]
can't we specify the SA as a parameter of the script (we often do that in legacy)

**Comment by andrerom** [comment]
Re SA, there is a global SA param for all Symfony commands, is this not working?

**Comment by lolautruche** [comment]
@pspanja Using `--siteaccess` is the best option by far. It should be possible to do this automatically, but complexity would be important compared to the benefit you'd get.

**Comment by lolautruche** [comment]
+1

**Comment by pspanja** [comment]
Thanks, I don't know how I missed it :blush: 
I guess this is ready for merge and backport then.

**Comment by andrerom** [comment]
+1

**Pull request merged**